### PR TITLE
Fix order of using directives should reset at preprocessor directives

### LINKED ISSUE #1161: Order of using directives should reset at preprocessor directives
The following block results in warnings regarding the order of preprocessor directives:

```csharp
using System;
using Microsoft.VisualStudio;
using MyList = System.Collections.Generic.List<int>;

#if FORVS10 || FORVS12 || FORVS13 || FORVS15
using Microsoft.CodeAnalysis;
#else
using Microsoft.CodeAnalysis;
#endif
```

However, each time a preprocessor line is reached, the relative order of items should be reset. In other words, two `using` directives should not be compared to each other if a preprocessor line exists between them.

## Actual Result

Warnings SA1210 and SA1209 are reported.

## Expected result

No warnings.

## Impact

This affects the usability of SA1208, SA1209, SA1210, SA1211, SA1216, and SA1217.

### PR DESCRIPTION
**Pull request by Noryoko**
- [x] Add regression tests for #1161
- [x] Fixes #1161

### CODE REVIEW

**Comment by Noryoko** [comment]
@sharwell Please review the tests for accuracy.

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/UsingDirectiveSyntaxHelpers.cs [hunk 1]
```diff
@@ -20,6 +20,44 @@ internal static class UsingDirectiveSyntaxHelpers
...
+        internal static bool IsPrecededByPreprocessorDirective(this UsingDirectiveSyntax usingDirective)
+        {
+            if (!usingDirective.HasLeadingTrivia)
+            {
+                return false;
+            }
+
+            foreach (var trivia in usingDirective.GetLeadingTrivia())
+            {
+                switch (trivia.Kind())
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/UsingDirectiveSyntaxHelpers.cs]: :bulb:

```csharp
if (trivia.IsDirective)
  return true;
```

**Comment by sharwell** [comment]
:question: Can you add tests for SA1208, SA1211, SA1216, and SA1217 as well?

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/UsingDirectiveSyntaxHelpers.cs (hunk 1 above)
Comment (Noryoko) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/UsingDirectiveSyntaxHelpers.cs]: That certainly simplifies things.

**Comment by Noryoko** [comment]
Will do.

**Pull request merged**